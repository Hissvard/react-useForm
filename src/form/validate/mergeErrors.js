const mergeArray = (fields, form) => {
  if (!fields) return form

  if (!Array.isArray(form)) {
    fields.error = form
    return fields
  }
  if (!Array.isArray(fields)) {
    form.error = fields
    return form
  }

  return form.reduce((acc, value, index) => {
    acc[index] = mergeObject(acc[index], value)

    return acc
  }, fields)
}

const mergeObject = (fields, form) => {
  if (!fields) return form
  if (!form) return fields

  return Object.entries(form).reduce((acc, [key, value]) => {

    if (Array.isArray(acc[key]) || Array.isArray(value)) {
      acc[key] = mergeArray(acc[key], value)
    } else {
      acc[key] = value
    }

    return acc
  }, fields)
}

export const mergeErrors = (fields, form) => {
  return mergeObject(fields, form)
}
