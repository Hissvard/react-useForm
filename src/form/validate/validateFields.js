import { Map } from 'immutable'

const validateItems = (items, getAllValues) => {
  return items.map(item => {
    return validateMap(item.get('fields'), getAllValues)
  })
}

const validateMap = (fields, getAllValues) => {
  return fields.entrySeq().reduce((acc, [key, field]) => {
    const value = field.getIn(['current', 'value'])
    const optional = field.getIn(['initial', 'optional'], false)
    const { validate, requiredMessage, name } = field.getIn(['initial', 'field'], {})

    const items = field.get('items')
    if (items) {
      return acc.set(key, validateItems(items), getAllValues)
    }

    if (validate) {
      const error = validate(value, name, getAllValues)
      if (error) {
        return acc.set(key, error)
      }
    }

    if (!optional && value !== false && !value) {
      return acc.set(key, requiredMessage || 'Required')
    }

    return acc
  }, Map()).toJS()
}

export const validateFields = (state) => {
  const getAllValues = state.get('getAllValues')
  return validateMap(state.get('fields'), getAllValues)
}
