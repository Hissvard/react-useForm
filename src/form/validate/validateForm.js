import { Map } from 'immutable'
import { getFieldValues } from '../helpers/getFieldValues'

export const validateForm = (validateFunctions, fieldData) => {
  if (!validateFunctions) return {}

  const values = getFieldValues(fieldData)
  let results = Map()
  let validators = validateFunctions
  if (!Array.isArray(validateFunctions)) { validators = [validateFunctions] }

  for (let v of validators) {
    results = results.mergeDeep(v(values))
  }
  return results.toJS()
}
