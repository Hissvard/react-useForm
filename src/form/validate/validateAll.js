import { validateForm } from './validateForm'
import { validateFields } from './validateFields'
import { mergeErrors } from './mergeErrors'

export const validateAll = (state, fieldData, validateFunctions, getAllValues) => {
  const formErrors = validateForm(validateFunctions, fieldData)
  const fieldErrors = validateFields(state, getAllValues)

  return mergeErrors(fieldErrors, formErrors)
  // dispatch(actions.validateAll(results))
}
