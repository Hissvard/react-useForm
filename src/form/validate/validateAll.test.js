import { validateForm } from './validateForm'
import { validateFields } from './validateFields'
import { validateAll } from './validateAll'
import { mergeErrors } from './mergeErrors'

jest.mock('./validateForm')
jest.mock('./validateFields')
jest.mock('./mergeErrors')

let state, fieldData, validate
beforeEach(() => {
  state = 'the state'
  fieldData = 'the field data'
  validate = jest.fn()

  const formErrors = {
    nickname: 'nickname, error',
    colors: [
      { color: 'color 0 error' },
    ],
  }
  const fieldErrors = {
    colors: [
      null,
      { color: 'color 1 error' },
    ],
  }

  validateForm.mockReturnValue(formErrors)
  validateFields.mockReturnValue(fieldErrors)
  mergeErrors.mockReturnValue('mergedErrors')
})

test('merges form and field errors, dispatches results', () => {
  expect(validateAll(state, fieldData, validate)).toEqual('mergedErrors')
})
