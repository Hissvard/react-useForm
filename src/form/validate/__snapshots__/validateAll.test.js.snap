// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merges form and field errors, dispatches results 1`] = `
Array [
  [MockFunction],
  "the field data",
]
`;

exports[`merges form and field errors, dispatches results 2`] = `
Array [
  "the state",
  "i will get all values",
]
`;
